File: swift-bridge/Sources/SceneKitBridge/Camera.swift
import SceneKit

@_cdecl("scn_camera_new")
public func scn_camera_new() -> UnsafeMutableRawPointer? {
    scnRetain(SCNCamera())
}

@_cdecl("scn_camera_get_field_of_view")
public func scn_camera_get_field_of_view(_ cameraHandle: UnsafeMutableRawPointer?) -> Double {
    guard let camera: SCNCamera = scnBorrow(cameraHandle) else { return 0 }
    return camera.fieldOfView
}

@_cdecl("scn_camera_set_field_of_view")
public func scn_camera_set_field_of_view(_ cameraHandle: UnsafeMutableRawPointer?, _ fieldOfView: Double) {
    guard let camera: SCNCamera = scnBorrow(cameraHandle) else { return }
    camera.fieldOfView = fieldOfView
}

@_cdecl("scn_camera_get_z_near")
public func scn_camera_get_z_near(_ cameraHandle: UnsafeMutableRawPointer?) -> Double {
    guard let camera: SCNCamera = scnBorrow(cameraHandle) else { return 0 }
    return camera.zNear
}

@_cdecl("scn_camera_set_z_near")
public func scn_camera_set_z_near(_ cameraHandle: UnsafeMutableRawPointer?, _ zNear: Double) {
    guard let camera: SCNCamera = scnBorrow(cameraHandle) else { return }
    camera.zNear = zNear
}

@_cdecl("scn_camera_get_z_far")
public func scn_camera_get_z_far(_ cameraHandle: UnsafeMutableRawPointer?) -> Double {
    guard let camera: SCNCamera = scnBorrow(cameraHandle) else { return 0 }
    return camera.zFar
}

@_cdecl("scn_camera_set_z_far")
public func scn_camera_set_z_far(_ cameraHandle: UnsafeMutableRawPointer?, _ zFar: Double) {
    guard let camera: SCNCamera = scnBorrow(cameraHandle) else { return }
    camera.zFar = zFar
}

@_cdecl("scn_camera_get_projection_transform")
public func scn_camera_get_projection_transform(
    _ cameraHandle: UnsafeMutableRawPointer?,
    _ outMatrix: UnsafeMutableRawPointer?
) -> Bool {
    guard let camera: SCNCamera = scnBorrow(cameraHandle) else { return false }
    return scnWriteMatrix4(camera.projectionTransform, out: outMatrix)
}

@_cdecl("scn_camera_set_projection_transform")
public func scn_camera_set_projection_transform(
    _ cameraHandle: UnsafeMutableRawPointer?,
    _ matrix: UnsafeMutableRawPointer?
) {
    guard let camera: SCNCamera = scnBorrow(cameraHandle),
          let matrix = scnReadMatrix4(matrix)
    else { return }
    camera.projectionTransform = matrix
}

File: src/camera.rs
use crate::ffi;
use crate::math::Matrix4;
use crate::private::handle_type;

handle_type!(Camera);

impl Camera {
    /// Creates a wrapped `SCNCamera` instance.
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_camera_new()) }
    }

    /// Mirrors `SCNCamera.fieldOfView`.
    #[must_use]
    pub fn field_of_view(&self) -> f64 {
        unsafe { ffi::scn_camera_get_field_of_view(self.ptr) }
    }

    /// Sets the `SCNCamera.fieldOfView` member.
    pub fn set_field_of_view(&self, field_of_view: f64) {
        unsafe { ffi::scn_camera_set_field_of_view(self.ptr, field_of_view) };
    }

    /// Mirrors `SCNCamera.zNear`.
    #[must_use]
    pub fn z_near(&self) -> f64 {
        unsafe { ffi::scn_camera_get_z_near(self.ptr) }
    }

    /// Sets the `SCNCamera.zNear` member.
    pub fn set_z_near(&self, z_near: f64) {
        unsafe { ffi::scn_camera_set_z_near(self.ptr, z_near) };
    }

    /// Mirrors `SCNCamera.zFar`.
    #[must_use]
    pub fn z_far(&self) -> f64 {
        unsafe { ffi::scn_camera_get_z_far(self.ptr) }
    }

    /// Sets the `SCNCamera.zFar` member.
    pub fn set_z_far(&self, z_far: f64) {
        unsafe { ffi::scn_camera_set_z_far(self.ptr, z_far) };
    }

    /// Mirrors `SCNCamera.projectionMatrix`.
    #[must_use]
    pub fn projection_matrix(&self) -> Matrix4 {
        let mut matrix = Matrix4::default();
        let _ = unsafe {
            ffi::scn_camera_get_projection_transform(self.ptr, matrix.as_mut_ptr().cast())
        };
        matrix
    }

    /// Sets the `SCNCamera.projectionMatrix` member.
    pub fn set_projection_matrix(&self, matrix: Matrix4) {
        unsafe {
            ffi::scn_camera_set_projection_transform(self.ptr, matrix.as_ptr().cast_mut().cast());
        };
    }
}
